game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+                    attackingUnits,
+                    false,
+                    parameters.getData(),
+                    battleSite,
+                    territoryEffects,
+                    parameters.isAmphibious,
+                    parameters.amphibiousLandAttackers),
+                parameters.getData());
+        final int defensePower =
+            DiceRoll.getTotalPower(
+                DiceRoll.getUnitPowerAndRollsForNormalBattles(
+                    defendingUnits,
+                    attackingUnits,
+                    defendingUnits,
+                    true,
+                    parameters.getData(),
+                    battleSite,
+                    territoryEffects,
+                    parameters.isAmphibious,
+                    parameters.amphibiousLandAttackers),
+                parameters.getData());
+        if (attackPower == 0 && defensePower == 0) {
+          tieProbability = 1;
+          hasResult = true;
+        }
+         */
+    }
+    // if a result was found, but no one is still alive, treat it as a draw
+    // see BattleResults::draw
+    if (hasResult && units.noMoreEnemies() && units.noMoreFriendlies()) {
+      loseProbability = 0;
+      winProbability = 0;
+      badProbability = 0;
+      tieProbability = 1;
+    }
+  }
+
+  /**
+   * Determines the result of the battle if it is finished
+   *
+   * @param units The two sides of the battle
+   * @return did the battle end
+   */
+  private boolean checkEndOfBranch(final StepUnits units) {
+    checkEndOfBattle(units);
+    if (hasResult) {
+      averageUnits = units;
+      averageUnits.recordChances();
+      averageRounds = round + 1;
+      return true;
+    }
+
+    // see if we've seen this situation and if so, grab the results
+    // if we've seen the situation but it doesn't yet have results
+    // then we are recursing.
+    final BattleStep cachedBattleStep = nodeCache.get(units);
+    if (cachedBattleStep != null) {
+      if (cachedBattleStep.hasResult) {
+        this.winProbability = cachedBattleStep.winProbability;
+        this.loseProbability = cachedBattleStep.loseProbability;
+        this.tieProbability = cachedBattleStep.tieProbability;
+        this.badProbability = cachedBattleStep.badProbability;
+        this.averageUnits = cachedBattleStep.averageUnits;
+        this.averageRounds = round + 1;
+        // System.out.println("Cached: " + this);
+      } else {
+        // this node reaches itself
+        // an example scenario would be two units fighting
+        // each round, there is a possibility that no one will hit
+        // so the units never change
+
+        final double[] siblingResults = new double[5];
+        siblingResults[0] = cachedBattleStep.winProbability;
+        siblingResults[1] = cachedBattleStep.loseProbability;
+        siblingResults[2] = cachedBattleStep.tieProbability;
+        siblingResults[3] = cachedBattleStep.averageRounds;
+        siblingResults[4] = 1.0;
```

Comment:
Instead of using a double array, why not create a dedicated type to hold all of these values?

Reply:
Good point.